src/Linux/local_cache.cpp

```diff
@@ -0,0 +1,367 @@
+
+    void throw_on_error() { this->should_throw_on_error = true; }
+
+    void open(const std::string& name, int flags, mode_t mode = 0)
+    {
+        RETRY_ON_EINTR(this->fd = ::open(name.c_str(), flags, mode), "open");
+
+        if (this->fd != -1)
+        {
+            const int lock_op = (flags & (O_WRONLY | O_RDWR)) ? LOCK_EX : LOCK_SH;
+            RETRY_ON_EINTR(::flock(this->fd, lock_op), "flock");
+        }
+    }
+
+    void close()
+    {
+        if (this->fd != -1)
+        {
+            int rc = ::close(this->fd);
+            this->fd = -1;
+
+            if (-1 == rc)
+            {
+                this->fail("Error closing file");
+            }
+        }
+    }
+
+    void truncate()
+    {
+        RETRY_ON_EINTR(::ftruncate(this->fd, 0), "ftruncate");
+    }
+
+    void read(void* data, size_t data_size)
+    {
+        RETRY_ON_EINTR(::read(this->fd, data, data_size), "read");
+    }
+
+    void write(const void* data, size_t data_size)
+    {
+        RETRY_ON_EINTR(::write(this->fd, data, data_size), "write");
+    }
+
+    off_t seek(off_t offset, int whence)
+    {
+        off_t rc = ::lseek(this->fd, offset, whence);
+
+        if (rc == -1)
+        {
+            this->fail("Error seeking file");
+        }
+
+        return rc;
+    }
+
+private:
+    int fd = -1;
+    bool has_failed = false;
+    bool should_throw_on_error = false;
+    std::string error_description;
+
+    void fail(const std::string& description)
+    {
+        this->has_failed = true;
+        this->error_description = description;
+
+        if (this->should_throw_on_error)
+        {
+            throw_errno(description);
+        }
+    }
+
+#undef RETRY_ON_EINTR
+};
+
+static std::string get_user_name()
+{
+    // sysconf could return -1, in which case we have to guess at the max size
+    size_t buf_size = std::max(sysconf(_SC_GETPW_R_SIZE_MAX), 128L);
+
+    int rc = 0;
+    do
+    {
+        std::vector<char> buf(buf_size);
+        passwd pwd{};
+        passwd* result = nullptr;
+        rc = getpwuid_r(geteuid(), &pwd, buf.data(), buf.size(), &result);
+
+        if (rc == 0)
+        {
+            return result->pw_name;
+        }
+        else if (rc == ERANGE)
+        {
+            buf_size *= 2;
+        }
+    } while (rc == ERANGE || rc == EINTR);
+
+    throw_errno("Error getting user name", rc);
+    return "";
+}
+
+static void make_dir(const std::string& dirname, mode_t mode)
+{
+    struct stat buf{};
+    int rc = stat(dirname.c_str(), &buf);
+    if (rc == 0)
+    {
+        if (S_ISDIR(buf.st_mode))
+        {
+            return;
+        }
+
+        throw std::runtime_error(dirname + " already exists, and is not a directory.");
+    }
+
+    rc = mkdir(dirname.c_str(), mode);
+    if (rc != 0)
+    {
+        throw_errno("Error creating directory '" + dirname + "'");
+    }
+}
+
+static void init_callback()
+{
+    std::string dirname = "/var/tmp/az-dcap-cache/";

```

Comment:
This might be a problematic location, because on my system I didn't have permission to write to it as a non-root user

Reply:
Sure, whats a better place to keep the cache root?